ludwig/trainers/trainer_llm.py

```diff
@@ -471,6 +473,65 @@ def evaluation(self, dataset, dataset_name, metrics_log, batch_size, progress_tr
         progress_tracker.llm_eval_examples = llm_eval_examples
         return append_metrics(self.model, dataset_name, metrics, metrics_log, progress_tracker)
 
+    def tune_batch_size(
+        self,
+        config: ModelConfigDict,
+        training_set: Dataset,
+        random_seed: int = default_random_seed,
+        max_trials: int = 20,
+        halving_limit: int = 3,
+        snapshot_weights: bool = True,
+        on_best_batch_size_updated: Optional[Callable[[int, float, int], None]] = None,
+        tune_for_training: bool = True,
+        max_sequence_length: Optional[int] = None,
+    ) -> int:
+        if max_sequence_length is None:
+            max_sequence_length = self.model.global_max_sequence_length
+        return super().tune_batch_size(
+            config,
+            training_set,
+            random_seed,
+            max_trials,
+            halving_limit,
+            snapshot_weights,
+            on_best_batch_size_updated,
+            tune_for_training,
+            max_sequence_length,
+        )
+
+    def _create_batch_size_evaluator(self) -> BatchSizeEvaluator:
+        trainer = self
+
+        class _TrainerBatchSizeEvaluator(BatchSizeEvaluator):
+            def reset(self):
+                trainer.model.reset_metrics()
+                trainer.optimizer.zero_grad()
+
+            def step(self, batch_size: int, max_sequence_length: Optional[int] = None):
+                # max_sequence_length here is the smaller value between the global max sequence length of the model and the model's context length
+                trainer.distributed.set_batch_size(trainer.dist_model, batch_size)
+                input_feature_name, input_feature = trainer.model.input_features.items()[0]
+                output_feature_name, output_feature = trainer.model.output_features.items()[0]
+                input_msl = input_feature.input_shape[0]
+                output_msl = output_feature.output_shape[0]
```

Comment:
Is this the most reliable way to get the MSL? Or should we be looking up properties in the feature's preprocessing configuration?

Reply:
Based on what I see [here](https://github.com/ludwig-ai/ludwig/blob/56e42c0573612fc1781555043269cda13d6bd835/ludwig/features/text_feature.py#L274), it looks like at least the input_shape should provide a tighter upper bound than looking at the preprocessing configuration, but maybe output_shape needs to be changed.

Reply:
Ok! Can you leave a quick comment about this in the code?